File: Colivi-frontend/src/features/housing/api/accommodationReviewService.ts
import api from '../../../lib/api';
import type {
  CreateReviewRequest,
  ReviewResponse,
  ReviewSummaryResponse,
  ReviewEligibilityResponse,
  PaginatedReviews,
} from '../types/review.types';

export const accommodationReviewService = {
  /**
   * GET /api/v1/listings/:id/reviews
   * Retrieves paginated reviews for a listing.
   */
  getListingReviews: async (
    listingId: string,
    page = 0,
    size = 10,
  ): Promise<PaginatedReviews> => {
    const response = await api.get<PaginatedReviews>(
      `/listings/${listingId}/reviews`,
      { params: { page, size } },
    );
    return response.data;
  },

  /**
   * GET /api/v1/listings/:id/reviews/summary
   * Returns average rating, total count and rating breakdown.
   */
  getListingReviewSummary: async (
    listingId: string,
  ): Promise<ReviewSummaryResponse> => {
    const response = await api.get<ReviewSummaryResponse>(
      `/listings/${listingId}/reviews/summary`,
    );
    return response.data;
  },

  /**
   * GET /api/v1/listings/:id/reviews/eligibility
   * Checks if current authenticated user has a CONFIRMED booking and hasn't reviewed yet.
   */
  checkEligibility: async (
    listingId: string,
  ): Promise<ReviewEligibilityResponse> => {
    const response = await api.get<ReviewEligibilityResponse>(
      `/listings/${listingId}/reviews/eligibility`,
    );
    return response.data;
  },

  /**
   * POST /api/v1/listings/:id/reviews
   * Submits a verified stay review.
   */
  createReview: async (
    listingId: string,
    payload: CreateReviewRequest,
  ): Promise<ReviewResponse> => {
    const response = await api.post<ReviewResponse>(
      `/listings/${listingId}/reviews`,
      payload,
    );
    return response.data;
  },

  /**
   * DELETE /api/v1/reviews/:id
   * Removes author's own review.
   */
  deleteReview: async (reviewId: string): Promise<void> => {
    await api.delete(`/reviews/${reviewId}`);
  },
};

File: Colivi-backend/src/main/java/com/vvu981/colivibackend/features/accommodation/service/AccommodationReviewService.java
package com.vvu981.colivibackend.features.accommodation.service;

import com.vvu981.colivibackend.features.accommodation.dto.CreateReviewRequest;
import com.vvu981.colivibackend.features.accommodation.dto.ReviewEligibilityResponse;
import com.vvu981.colivibackend.features.accommodation.dto.ReviewResponse;
import com.vvu981.colivibackend.features.accommodation.dto.ReviewSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface AccommodationReviewService {

    ReviewResponse createReview(UUID listingId, CreateReviewRequest request, UUID currentUserId);

    Page<ReviewResponse> getListingReviews(UUID listingId, Pageable pageable);

    ReviewSummaryResponse getListingReviewSummary(UUID listingId);

    ReviewEligibilityResponse checkEligibility(UUID listingId, UUID currentUserId);

    List<ReviewResponse> getReviewsByCity(String city);

    void deleteReview(UUID reviewId, UUID currentUserId, boolean isAdmin);
}
